src/Features/VisualBasic/Portable/ChangeSignature/VisualBasicChangeSignatureService.vb

```diff
@@ -407,32 +423,44 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ChangeSignature
         Private Function PermuteArgumentList(
             arguments As SeparatedSyntaxList(Of ArgumentSyntax),
             permutedSignature As SignatureChange,
-            document As Document,
             declarationSymbol As ISymbol,
             Optional isReducedExtensionMethod As Boolean = False) As SeparatedSyntaxList(Of ArgumentSyntax)
 
-            Dim newArguments As List(Of IUnifiedArgumentSyntax) = MyBase.PermuteArguments(document, declarationSymbol, arguments.Select(Function(a) UnifiedArgumentSyntax.Create(a)).ToList(), permutedSignature, isReducedExtensionMethod)
+            Dim newArguments As List(Of IUnifiedArgumentSyntax) = MyBase.PermuteArguments(
+                declarationSymbol, arguments.Select(Function(a) UnifiedArgumentSyntax.Create(a)).ToList(), permutedSignature,
+                Function(callsiteValue) UnifiedArgumentSyntax.Create(SyntaxFactory.SimpleArgument(SyntaxFactory.ParseExpression(callsiteValue))),
+                isReducedExtensionMethod)
 
             Dim numSeparatorsToSkip = arguments.Count - newArguments.Count
             Return SyntaxFactory.SeparatedList(newArguments.Select(Function(a) CType(DirectCast(a, UnifiedArgumentSyntax), ArgumentSyntax)), GetSeparators(arguments, numSeparatorsToSkip))
         End Function
 
-        Private Function PermuteDeclaration(Of T As SyntaxNode)(list As SeparatedSyntaxList(Of T), updatedSignature As SignatureChange) As SeparatedSyntaxList(Of T)
-            Dim originalParameters = updatedSignature.OriginalConfiguration.ToListOfParameters()
-            Dim reorderedParameters = updatedSignature.UpdatedConfiguration.ToListOfParameters()
+        Private Function PermuteDeclaration(Of T As SyntaxNode)(
+                                                               parameterList As SeparatedSyntaxList(Of T),
+                                                               updatedSignature As SignatureChange,
+                                                               createNewParameterMethod As Func(Of AddedParameter, T)) As SeparatedSyntaxList(Of T)
+            Dim basePermuteDeclaration = PermuteDeclarationBase(parameterList, updatedSignature, createNewParameterMethod)
+            Return SyntaxFactory.SeparatedList(basePermuteDeclaration.Item1, basePermuteDeclaration.Item2)
+        End Function
 
-            Dim newParameters = New List(Of T)
-            For Each newParam In reorderedParameters
-                Dim pos = originalParameters.IndexOf(newParam)
-                Dim param = list(pos)
-                newParameters.Add(param)
-            Next
+        Private Shared Function CreateNewParameterSyntax(addedParameter As AddedParameter) As ParameterSyntax
+            Return CreateNewParameterSyntax(addedParameter, skipType:=False)
+        End Function
+
+        Private Shared Function CreateNewParameterSyntax(addedParameter As AddedParameter, skipType As Boolean) As ParameterSyntax
+            Return SyntaxFactory.Parameter(
```

Comment:
why can't we use SyntaxGenerator or ICodeGenerationService for this sort of thing?

Reply:
https://github.com/dotnet/roslyn/issues/41130